instance/tests/models/test_instance.py

```diff
@@ -552,3 +555,61 @@ def test_status_transitions(self):
         instance_provisioning_failed._status_to_configuration_failed()
         self.assertEqual(instance_provisioning_failed.status, Instance.Status.ConfigurationFailed)
         self._assert_status_conditions(instance_provisioning_failed, is_healthy_state=False)
+
+        # Reprovisioning from steady states
+        steady_states = (
+            Instance.Status.New,
+            Instance.Status.Error,
+            Instance.Status.ConfigurationFailed,
+            Instance.Status.Running,
+            Instance.Status.Terminated
+        )
+        for steady_state in steady_states:
+            instance = SingleVMOpenEdXInstanceFactory(status=steady_state)
+            self.assertEqual(instance.status, steady_state)
+            instance._status_to_waiting_for_server()
+            self.assertEqual(instance.status, Instance.Status.WaitingForServer)
+            self._assert_status_conditions(instance, is_steady_state=False)
+
+    @data(
+        {
+            'name': '_status_to_waiting_for_server',
+            'from_states': [
+                Instance.Status.New,
+                Instance.Status.Error,
+                Instance.Status.ConfigurationFailed,
+                Instance.Status.Running,
+                Instance.Status.Terminated,
+            ],
+        },
+        {
+            'name': '_status_to_error',
+            'from_states': [Instance.Status.WaitingForServer],
+        },
+        {
+            'name': '_status_to_configuring_server',
+            'from_states': [Instance.Status.WaitingForServer],
+        },
+        {
+            'name': '_status_to_configuration_failed',
+            'from_states': [Instance.Status.ConfiguringServer],
+        },
+        {
+            'name': '_status_to_running',
+            'from_states': [Instance.Status.ConfiguringServer],
+        },
+        {
+            'name': '_status_to_terminated',
+            'from_states': [Instance.Status.Running],
+        },
+    )
+    def test_invalid_status_transitions(self, transition):
+        """
+        Test that invalid status transitions raise exception
+        """
+        invalid_from_states = (state for state in Instance.Status.states if state not in transition['from_states'])
+        for invalid_from_state in invalid_from_states:
+            instance = SingleVMOpenEdXInstanceFactory(status=invalid_from_state)
+            self.assertEqual(instance.status, invalid_from_state)
+            with self.assertRaises(WrongStateException):
+                getattr(instance, transition['name'])()
```

Comment:
Comment: This test is fine, but this behavior should already be guaranteed and unit tested by the `ResourceState` code, and we shouldn't have to have separate tests that "invalid status transitions raise exception" every time we use `ResourceState`.